src/rules/LinearAlgebra/blas.jl

```diff
@@ -0,0 +1,138 @@
+    xtape = _safe_similar(x, n)
+    BLAS.blascopy!(n, x, incx, xtape, 1)
+    return xtape
+end
+
+function _maybe_primal_shadow(config, func, args)
+    needs_primal = EnzymeRules.needs_primal(config)
+    needs_shadow = EnzymeRules.needs_shadow(config)
+    if needs_primal || needs_shadow
+        r = func(args...)
+    else
+        r = nothing
+    end
+    primal = needs_primal ? r : nothing
+    shadow = needs_shadow ? zero(r) : nothing
+    return primal, shadow
+end
+
+_map_tuple(f, xs::Tuple...) = map(f, xs...)
+_map_tuple(f, xs...) = f(xs...)
+
+for (fname, Ttype) in ((:dot, :BlasReal), (:dotu, :BlasComplex), (:dotc, :BlasComplex))
+    @eval begin
+        function EnzymeRules.forward(
+            func::Const{typeof(BLAS.$fname)},
+            RT::Type{
+                <:Union{
+                    Const,DuplicatedNoNeed,Duplicated,BatchDuplicatedNoNeed,BatchDuplicated
+                },
+            },
+            n::Const{<:Integer},
+            X::ConstOrBatchDuplicated{<:Union{Ptr{T},AbstractArray{T}}},
+            incx::Const{<:Integer},
+            Y::ConstOrBatchDuplicated{<:Union{Ptr{T},AbstractArray{T}}},
+            incy::Const{<:Integer},
+        ) where {T<:BLAS.$Ttype}
+            RT <: Const && return func.val(n.val, X.val, incx.val, Y.val, incy.val)
+
+            dval = if !(X isa Const) && !(Y isa Const)
+                _map_tuple(X.dval, Y.dval) do dX, dY
+                    func.val(n.val, dX, incx.val, Y.val, incy.val) +
+                    func.val(n.val, X.val, incx.val, dY, incy.val)
+                end
+            elseif !(X isa Const)
+                _map_tuple(dX -> func.val(n.val, dX, incx.val, Y.val, incy.val), X.dval)
+            elseif !(Y isa Const)
+                _map_tuple(dY -> func.val(n.val, X.val, incx.val, dY, incy.val), Y.dval)
+            else
+                zero(T)
+            end
+
+            if RT <: Union{DuplicatedNoNeed,BatchDuplicatedNoNeed}
+                return dval
+            else
+                val = func.val(n.val, X.val, incx.val, Y.val, incy.val)
+                return RT(val, dval)
+            end
+        end
+
+        function EnzymeRules.augmented_primal(
+            config::EnzymeRules.ConfigWidth{1},
+            func::Const{typeof(BLAS.$fname)},
+            RT::Type{<:Union{Const,Active}},
+            n::Const{<:Integer},
+            X::ConstOrDuplicated{<:Union{Ptr{T},AbstractArray{T}}},
+            incx::Const{<:Integer},
+            Y::ConstOrDuplicated{<:Union{Ptr{T},AbstractArray{T}}},
+            incy::Const{<:Integer},
+        ) where {T<:BLAS.$Ttype}
+            primal, shadow = _maybe_primal_shadow(
+                config, func.val, (n.val, X.val, incx.val, Y.val, incy.val)
+            )
+
+            # build tape
+            if !(RT <: Const)
+                _, _, Xow, _, Yow = EnzymeRules.overwritten(config)
+                if Xow || BLAS.$fname === BLAS.dotu
```

Comment:
An alternative to always taping for `dotu` would be to to do what we do hear when `X` or `Y` is overwritten but otherwise broadcast instead of using `axpy!` in `reverse` to include the conjugation operation.